[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/gamayos/finite-ring-space/blob/main/src/28-flavour/28-flavour-main.ipynb)

## Fermion Flavour Sector over Finite Relational Substrate (Akhtman & Voether, 2026)

**Validation Package.** `finite-ring-space/src/28-flavour` — the paper's sixteen validation scripts as written, run through one registry: a
family check is one script (`flv.<stem>`), its micro-checks the script's own verdict lines (twelve of the sixteen print one per check)
together with the registry's predicates, which decide the stated values explicitly where a script prints without asserting
(`framed_koide`, `theta13`, `scale_a`, `coupling_anchor`) and pin the headline numerals elsewhere; each family names the row(s) of the
paper's predicate ledger it witnesses (the paper's Appendix "Predicate ledger", rows cited as `28:XN`; public copy at
`docs/28-flavour/28-flavour-ledger.html`), and the ledger's source column cites the family ids in return. Master-ledger rows of the
corpus reached through the paper rows: `00:K2`, `00:K3`, `00:K4`, `00:J2`, `00:J4`, `00:D2`, `00:D8`, `00:Z6`, `00:Z8`, `00:C7`, `00:B5`.

**Kinds** — the suite's own three classes, recorded per family. `EXACT`: integer, F_p, F_{p²} or cyclotomic arithmetic throughout, no
float in any asserted claim. `MIXED`: the script carries its own exact check for at least one claim, the continuum confined to
labelled [approx] comparisons. `APPROX`: continuum/numerical by construction — a numerical reconfirmation of an identity proven
exactly in `exact_core`, a comparison with measured data, or an Ω-hard reading.

**Run.** Cell by cell, or *Runtime → Run all*; ≈ 1 minute on Colab (`pmns_cp` is the long one; `tm2_jointfit` needs scipy). The last
cell writes `results.json` and fails loudly if any family fails.

In [1]:
# --- environment: clone the package if this notebook is not already running inside it (Colab). numpy, sympy and scipy are present on Colab.
import os
if not os.path.exists("flvcommon.py"):
    if not os.path.exists("finite-ring-space"):
        !git clone --filter=blob:none --sparse https://github.com/gamayos/finite-ring-space.git
        os.chdir("finite-ring-space")
        !git sparse-checkout set src/28-flavour
        os.chdir("src/28-flavour")
    else:
        os.chdir("finite-ring-space/src/28-flavour")
print("working directory:", os.path.join(*os.getcwd().split(os.sep)[-3:]))

working directory: finite-ring-space/src/28-flavour


In [2]:
import importlib, flvcommon as gc
importlib.reload(gc)         # a fresh registry if the notebook is re-run
print("families registered:", len(gc.LABELS))

families registered: 16


## The float-free core
The Koide and circulant identities in Q(ω, √2) and over framed rationals in F_{p²}/F_p; the π/12 boundary; the cubic Gauss sums as integer group-ring identities; the trimaximal magic matrix with maximal Jarlskog; the amplitude diagonal N(1 − ζ_n) = 2, 3 as integers in F_p.

| id | script | kind | claim | ledger rows |
|---|---|---|---|---|
| `flv.exact_core` | `exact_core.py` | EXACT | the float-free core, 25 verifications: the circulant Koide identities in Q(ω, √2) (Σλ = 3a, Σλ² = 3a² + 6|b|², Q = 1/3 + (2/3)ρ², ρ² = 1/2 ⇒ Q = 2/3, Hermitian √M with exact eigenvalues); the π/12 boundary as rational π with the electron amplitude vanishing; the cubic Gauss-sum reality and the π/3 quantisation of the drive-twisted overlap as integer identities in Z[Z/3 × Z/p] for p = 17, 29, 41, 53; the trimaximal magic matrix with maximal Jarlskog J = 1/(6√3) in Q(ω); Koide over framed rationals in F_{p²}/F_p (Q = 2/3 exact, the quarter-turn the rational N(b)/a² = 1/2) and the amplitude diagonal N(1 − i) = 2, N(1 − ω) = 3 as integers in F_p for p = 17, 53, 89 | 28:C1, 28:C2, 28:C3, 28:C4, 28:C10, 28:Z1 |
| `flv.framed_koide` | `framed_koide.py` | EXACT | the Koide identity natively over framed rationals (F_{p²}/F_p, Frobenius conjugation, no cyclotomics) on p = 17, 53, 89: the √-mass eigenvalues framed rationals in F_p, Q = 2/3 exactly, the '√2 amplitude' the rational N(b)/a² = 1/2, ω the finite cube root | 28:C2, 28:X1 |

Ledger rows witnessed: 28:C1, 28:C2, 28:C3, 28:C4, 28:C10, 28:X1, 28:Z1.

In [3]:
for fam in ["flv.exact_core", "flv.framed_koide"]:
    print(f"\n— {gc.script_of(fam)}.py"); gc.run_block(fam)


— exact_core.py


A. Koide / circulant in the cyclotomic field Q(omega, sqrt2)
[EXACT] A.omega: omega^3=1 and 1+omega+omega^2=0 (exact)
[EXACT] A.sum1: sum_k lambda_k = 3*a = 3a
[EXACT] A.sum2: sum_k lambda_k^2 = 3a^2 + 6|b|^2  (exact; the omega-cross terms cancel)
[EXACT] A.Qform: Q = 1/3 + (2/3) rho^2  (exact)
[EXACT] A.koide: r = sqrt2  (rho=1/sqrt2, the quarter-turn diagonal)  =>  Q = 2/3 exactly
[EXACT] A.herm: sqrt(M)=a*1+b*P+conj(b)*P^2 is Hermitian (P^2=P^dagger)


[EXACT] A.eig: F^{-1} M F diagonal with entries a+b w^k+conj(b) w^{2k}  (exact eigenvalues)

B. The pi/12 boundary (exact rational-pi + algebraic)
[EXACT] B.pi12: delta_LO = 3pi/4 - 2pi/3 = pi/12 (exact)
[EXACT] B.zero: 1 + sqrt2*cos(3pi/4) = 1 + sqrt2*(-1/sqrt2) = 0 (electron massless at boundary)

C. Gauss sums: reality & pi/3-quantisation as INTEGER identities
[EXACT] C.real_p17: p=17: symmetric cubic overlap b = conj(b) EXACTLY (integer-real); no phase
[EXACT] C.real_p29: p=29: symmetric cubic overlap b = conj(b) EXACTLY (integer-real); no phase
[EXACT] C.real_p41: p=41: symmetric cubic overlap b = conj(b) EXACTLY (integer-real); no phase
[EXACT] C.real_p53: p=53: symmetric cubic overlap b = conj(b) EXACTLY (integer-real); no phase
[EXACT] C.cube_p17: p=17: drive-twisted b^3 = conj(b^3) EXACTLY  =>  arg(b) in (pi/3)Z
[EXACT] C.cube_p29: p=29: drive-twisted b^3 = conj(b^3) EXACTLY  =>  arg(b) in (pi/3)Z
[EXACT] C.cube_p41: p=41: drive-twisted b^3 = conj(b^3) EXACTLY  =>  arg(b) in (

[EXACT] E.framed_p53: p=53: Q=2/3 exact in F_p; sqrt-mass eigenvalues framed rationals [(1, 1), (-3, 5), (5, 6)]; omega=(-1, 2)+(-5, 2)t; quarter-turn = N(b)/a^2 = 1/2 (rational, not sqrt2)


[EXACT] E.framed_p89: p=89: Q=2/3 exact in F_p; sqrt-mass eigenvalues framed rationals [(5, 1), (1, 7), (2, 5)]; omega=(-1, 2)+(-4, 5)t; quarter-turn = N(b)/a^2 = 1/2 (rational, not sqrt2)

F. The amplitude diagonal over FRAMED RATIONALS: r^2 = N(1-zeta_n) = 2, 3 (integers in F_p)
[EXACT] F.diag_p17: p=17: N(1-i)=2 (=2: four-fold, r^2=2 -> Q=2/3); N(1-omega)=3 (=3: three-fold/colour, r^2=3 -> Q=5/6) -- integers in F_p


[EXACT] F.diag_p53: p=53: N(1-i)=2 (=2: four-fold, r^2=2 -> Q=2/3); N(1-omega)=3 (=3: three-fold/colour, r^2=3 -> Q=5/6) -- integers in F_p


[EXACT] F.diag_p89: p=89: N(1-i)=2 (=2: four-fold, r^2=2 -> Q=2/3); N(1-omega)=3 (=3: three-fold/colour, r^2=3 -> Q=5/6) -- integers in F_p

EXACT CORE: 25/25 float-free verifications hold
PASS the script's own tally: 25/25 float-free verifications hold
  [PASS] flv.exact_core         EXACT  [28:C1, 28:C2, 28:C3, 28:C4, 28:C10, 28:Z1] the float-free core, 25 verifications: the circulant Koide identities in Q(ω, √2) (Σλ = 3a  --  26 checks, 5.4 s

— framed_koide.py
Koide over framed rationals (F_{p^2}/F_p, Frobenius conjugation), no cyclotomics:
  p=17  (K=F_{17^2}, t^2=3, Frobenius t->-t)
    omega (cube root, order 3) = (8, 2) = -1/2 + 2 t
    a = 1 ;  b = (2) + (2) t ;  N(b)/a^2 = 1/2  (= 1/2, the quarter-turn, a RATIONAL)
    sqrt-mass eigenvalues (framed rationals in F_p): -2/3, 1/3, 1/2   in F_p: True
    Q = sum(lam^2)/(sum lam)^2 = 12  =  2/3   (2/3 = 12): EXACT 2/3


  p=53  (K=F_{53^2}, t^2=2, Frobenius t->-t)
    omega (cube root, order 3) = (26, 24) = -1/2 + -5/2 t
    a = 1 ;  b = (0) + (-1/7) t ;  N(b)/a^2 = 1/2  (= 1/2, the quarter-turn, a RATIONAL)
    sqrt-mass eigenvalues (framed rationals in F_p): 1, -3/5, 5/6   in F_p: True
    Q = sum(lam^2)/(sum lam)^2 = 36  =  2/3   (2/3 = 36): EXACT 2/3


  p=89  (K=F_{89^2}, t^2=3, Frobenius t->-t)
    omega (cube root, order 3) = (44, 17) = -1/2 + -4/5 t
    a = 1 ;  b = (2) + (4) t ;  N(b)/a^2 = 1/2  (= 1/2, the quarter-turn, a RATIONAL)
    sqrt-mass eigenvalues (framed rationals in F_p): 5, 1/7, 2/5   in F_p: True
    Q = sum(lam^2)/(sum lam)^2 = 60  =  2/3   (2/3 = 60): EXACT 2/3

  ALL shells: Q = 2/3 exact over framed rationals  ->  True
  the 'sqrt2 amplitude' is the framed rational N(b)/a^2 = 1/2 ; sqrt2 is its
  continuum reading.  omega is the finite cube root, not zeta_3 in C.
PASS Q = 2/3 exact over framed rationals on all three shells p = 17, 53, 89 (the script's verdict)
  p=17  (K=F_{17^2}, t^2=3, Frobenius t->-t)
    omega (cube root, order 3) = (8, 2) = -1/2 + 2 t
    a = 1 ;  b = (2) + (2) t ;  N(b)/a^2 = 1/2  (= 1/2, the quarter-turn, a RATIONAL)
    sqrt-mass eigenvalues (framed rationals in F_p): -2/3, 1/3, 1/2   in F_p: True
    Q = sum(lam^2)/(sum lam)^2 = 12  =  2/3   (2/3 = 12): EXACT 2/3
PASS p=17: Q = 2/3 in

  p=53  (K=F_{53^2}, t^2=2, Frobenius t->-t)
    omega (cube root, order 3) = (26, 24) = -1/2 + -5/2 t
    a = 1 ;  b = (0) + (-1/7) t ;  N(b)/a^2 = 1/2  (= 1/2, the quarter-turn, a RATIONAL)
    sqrt-mass eigenvalues (framed rationals in F_p): 1, -3/5, 5/6   in F_p: True
    Q = sum(lam^2)/(sum lam)^2 = 36  =  2/3   (2/3 = 36): EXACT 2/3
PASS p=53: Q = 2/3 in F_p with the √-mass eigenvalues in F_p (recomputed)


  p=89  (K=F_{89^2}, t^2=3, Frobenius t->-t)
    omega (cube root, order 3) = (44, 17) = -1/2 + -4/5 t
    a = 1 ;  b = (2) + (4) t ;  N(b)/a^2 = 1/2  (= 1/2, the quarter-turn, a RATIONAL)
    sqrt-mass eigenvalues (framed rationals in F_p): 5, 1/7, 2/5   in F_p: True
    Q = sum(lam^2)/(sum lam)^2 = 60  =  2/3   (2/3 = 60): EXACT 2/3
PASS p=89: Q = 2/3 in F_p with the √-mass eigenvalues in F_p (recomputed)
  [PASS] flv.framed_koide       EXACT  [28:C2, 28:X1] the Koide identity natively over framed rationals (F_{p²}/F_p, Frobenius conjugation, no c  --  4 checks, 3.7 s


## The charged sectors: Koide, the winding kernel, the phase lock
The Tier-A/B figures against data; the circulant winding kernel with δ_LO = π/12 and the per-sector extraction; the cross-sector phase lock 1 : 1/2 : 1/3 and the small-shell quantisation; the revision identities (J, δ_LO = 1/24, the signed Koide, the cube invariant) with their labelled confrontations.

| id | script | kind | claim | ledger rows |
|---|---|---|---|---|
| `flv.tier_b` | `tier_b.py` | MIXED | the Tier-A/B figures: the measured charged-lepton Q against 2/3 (6 × 10⁻⁶), the symbolic Σa = 3, Σa² = 3 + (3/2)r², Q(√2) = 2/3 for all δ and Q = 2/3 ⇔ r = √2; ρ = |â₁|/|â₀| = 1/√2 and the 45° democratic angle; the λ-power texture (down/lepton λ^{4,2,0}, up λ^{8,4,0}) and the Gatto V_us = √(m_d/m_s) (within 5 %); the Wolfenstein CKM powers; the Georgi–Jarlskog double ratio against N_c² = 9 (within 20 %) and b–τ; the lopsided M_e = M_dᵀ giving a large PMNS θ₂₃ with a small CKM θ₂₃, against a seeded hierarchical-seesaw control | 28:C2, 28:C6, 28:C8, 28:X3, 28:X5, 28:P1 |
| `flv.m10` | `m10.py` | MIXED | the winding kernel: the circulant √M Hermitian with eigenvalues a + 2|b| cos(δ + 2πk/3); Q = 1/3 + r²/6 symbolically and r = √2 ⇒ 2/3; δ_LO = 3π/4 − 2π/3 = π/12 with the electron massless at leading order and m_μ/m_τ = 0.072 against 0.060; the per-sector extraction (lepton r = √2 to 2 × 10⁻³, quark amplitudes above √2, r_u near √3); the physical δ off the π/12 grid; the parameter reduction | 28:C1, 28:C2, 28:C3, 28:X7 |
| `flv.delta` | `delta.py` | MIXED | the cross-sector phase lock δ_ℓ : δ_d : δ_u = 1 : 1/2 : 1/3 in the working scheme (medians over 20 000 seeded mass draws); the symmetric cubic overlap real on every p ≡ 5 (mod 12) below 170 and the drive-twisted overlap complex; small-shell phases quantised to multiples of π/3; δ_LO = π/12 | 28:C3, 28:C5, 28:Z1, 28:Z2 |
| `flv.revision_checks` | `revision_checks.py` | MIXED | the revision identities: J = Im ω/9 = 1/(6√3) and J² = 1/108; δ_LO = 3/8 − 1/3 = 1/24 cycle; the signed-amplitude Koide = 2/3 at every δ for r = √2; the cube-invariant determinant ∏√m = a³(cos 3δ/√2 − 1/2) [exact]; 3δ₀ = Q (δ₀ = Q/3), the positive-root neutrino Koide values, the normal-branch preference (> 20×), the Q_u scheme spread [approx] | 28:C3, 28:C10, 28:C11b, 28:P2, 28:Z1 |

Ledger rows witnessed: 28:C1, 28:C2, 28:C3, 28:C5, 28:C6, 28:C8, 28:C10, 28:C11b, 28:P1, 28:P2, 28:X3, 28:X5, 28:X7, 28:Z1, 28:Z2.

In [4]:
for fam in ["flv.tier_b", "flv.m10", "flv.delta", "flv.revision_checks"]:
    print(f"\n— {gc.script_of(fam)}.py"); gc.run_block(fam)


— tier_b.py
M8  KOIDE:  Q = 2/3 from cube-root generations + quarter-turn bisection
[PASS] M8.data: Q_lep = 0.666661  vs 2/3 = 0.666667  (dev 6.16e-06)
[PASS] M8.sum1: Sum a_i = 3 (delta-independent)
[PASS] M8.sum2: Sum a_i^2 = 3*r**2/2 + 3
[PASS] M8.exact23: Q(r=sqrt2) = 2/3  for ALL delta  -> exact 2/3
[PASS] M8.ampl: Q=2/3  <=>  r in [-sqrt(2), sqrt(2)]  (positive root sqrt2 = quarter-turn diagonal)
[PASS] M8.rho: rho = |a_1|/|a_0| = 0.707100  vs 1/sqrt2 = 0.707107
[PASS] M8.parseval: 1/3+2/3 rho^2 = 0.666661 reproduces Q_lep
[PASS] M8.angle45: angle(sqrt m, democratic) = 44.9997 deg  (cos^2=0.50000~1/2)
   [report] Koide phase delta = 0.22227 rad (2/9 = 0.22222) -- a Tier-C winding datum, reported not claimed

M5/M6  ROLE-DEPTH FN CHARGES + lambda-POWER TEXTURE
   down  : powers of lam = [4.56,2.55,0.00]  round=[5, 3, 0]  FN-pred=[4, 2, 0]
   lepton: powers of lam = [5.47,1.89,0.00]  round=[5, 2, 0]  FN-pred=[4, 2, 0]
   up    : powers of lam = [7.57,3.29,0.00]  round=[8, 3, 0]  F

   control: median CKM 13 deg, median seesaw-PMNS (hierarchical Dirac) 21 deg -> hierarchical Dirac mass does NOT enhance (the reported finding)
[PASS] M9.seesaw_ctrl: hierarchical-Dirac seesaw alone does not robustly enlarge angles (enhancement needs the lopsided M_e=M_d^T route, route A)

SUMMARY: 17/17 exact/figure-of-merit checks PASS
PASS the script's own tally: 17/17 checks
PASS [approx] measured Q_ℓ − 2/3 = 6.2e-06 < 1e-5 (the paper's 6 × 10⁻⁶)
PASS [approx] Gatto V_us = √(m_d/m_s) = 0.2236 against 0.2243 within 0.5 % (the paper's 0.3 %)
PASS [approx] Georgi–Jarlskog double ratio = 10.34 against 9: the open excess in 10–20 % (the paper's 15 %)
PASS [approx] lopsided: PMNS θ₂₃ = 52.5°, CKM θ₂₃ = 1.4° (the paper's 52.5° and 1.4°)
  [PASS] flv.tier_b             MIXED  [28:C2, 28:C6, 28:C8, 28:X3, 28:X5, 28:P1] the Tier-A/B figures: the measured charged-lepton Q against 2/3 (6 × 10⁻⁶), the symbolic Σ  --  22 checks, 0.4 s

— m10.py
1.  Circulant amplitude matrix from generation uni

[PASS] 1.robust_down: down/lepton median 0.495 in [0.474,0.516] (->1/2)
[PASS] 1.robust_up: up/lepton median 0.334 in [0.326,0.342] (->1/3)
[PASS] 1.reduce: four sector phases reduce to ONE: delta_0 = delta_lepton (~2/9 rad)
   scheme note: with the pole charm mass 1.67 GeV, delta_up/delta_lepton = 0.388 (working MSbar scheme: 0.335);
   the 1:1/2:1/3 lock is a statement in the MSbar working scheme (m_c(m_c)); Q_u=5/6 at pole masses is not simultaneously available.

(2)/(3)  delta is a drive-orientation phase; small shells quantise it
[PASS] 2.sym_real: symmetric cubic overlap is REAL for all p=5mod12 (max|Im|=6.6e-15); conjugation u->ubar cancels the phase
[PASS] 2.drive_complex: drive-twisted overlap is COMPLEX (max|Im|=12.27); the drive breaks conjugation -> delta is a parity-linked drive-orientation phase
[PASS] 3.quantised: small-shell phases are QUANTISED to multiples of pi/3 (max off-grid=6.0e-15); the smooth observed delta=12.7deg is NOT a small-shell phase -> open (large-Omega

  [PASS] B2.branch: [approx] r=sqrt2 fit 66x tighter for normal than inverted (boundary branch selects NO)
  [PASS] B3.Qu: [approx] Q_u: on-shell 0.832 ~ 5/6=0.833; MSbar(M_Z) 0.888; exact object r_u^2=3

SUMMARY: 9/9 checks PASS  (A exact framed-rational, B labelled [approx])
PASS the script's own tally: 9 checks, none failed (each asserts)
PASS [approx] 3δ₀ = 0.66666 = Q = 0.66666 within 1e-4 (δ₀ = Q/3 = 2/9)
PASS [approx] on-shell Q_u = 0.832 = 5/6 within 0.005; MS̄(M_Z) Q_u = 0.888 > 0.86
  [PASS] flv.revision_checks    MIXED  [28:C3, 28:C10, 28:C11b, 28:P2, 28:Z1] the revision identities: J = Im ω/9 = 1/(6√3) and J² = 1/108; δ_LO = 3/8 − 1/3 = 1/24 cycl  --  12 checks, 0.8 s


## The quark sector: amplitudes, doubling, the spurion, the reactor angle
The colour dressing of the quark amplitudes (r_u ≈ √3, Q_u = 5/6 in framed arithmetic); the up-sector doubling from 10·10; the Cabibbo spurion folded onto the winding kernel; θ₁₃ = θ_C/√2 as a leading-order estimate.

| id | script | kind | claim | ledger rows |
|---|---|---|---|---|
| `flv.quark_amp` | `quark_amp.py` | MIXED | the quark amplitudes: the Georgi–Jarlskog colour dressing shifts Q from 2/3 to 0.745 against the observed Q_d = 0.731; r_u = 1.759 against √3 (the cube-root colour diagonal, robust over the mass ranges); r² = N(1 − ζ_n) = 2, 3 as integers in F_p for p = 17, 53, 89; Q_u = 0.849 in the working scheme, 0.832 with the pole charm mass (the scheme band of 5/6) | 28:B4, 28:C4, 28:X2, 28:P2 |
| `flv.up_doubling` | `up_doubling.py` | MIXED | the up-sector doubling: the effective Froggatt–Nielsen powers (8, 4) up against (4, 2) down and lepton, up = 2 × down at the integer level; the 10·10 against 10·5̄ structural map with the 10-charge a = (4, 2, 0); the winding self-pair reading | 28:B3, 28:C7, 28:X6 |
| `flv.spurion` | `spurion.py` | APPROX | the Cabibbo spurion: the down λ-powers from the role charges, the Gatto V_us = √(m_d/m_s) within 2 %, λ from the down circulant (δ₀/2, r_d) within 0.03, λ ≈ δ₀ ≈ 2/9 within 2 % [approx, data] | 28:C6, 28:X4, 28:Z1 |
| `flv.theta13` | `theta13.py` | APPROX | the reactor angle as a leading-order estimate: sin θ₁₃ = sin θ_C/√2 = 0.159 against 0.149 (7 %), quark–lepton complementarity θ₁₂ + θ_C ≈ 45°, the fold onto δ₀/√2, sin θ₂₃ ≈ 1/√2 — four tolerances [approx, data] | 28:C9, 28:X9 |

Ledger rows witnessed: 28:B3, 28:B4, 28:C4, 28:C6, 28:C7, 28:C9, 28:P2, 28:X2, 28:X4, 28:X6, 28:X9, 28:Z1.

In [5]:
for fam in ["flv.quark_amp", "flv.up_doubling", "flv.spurion", "flv.theta13"]:
    print(f"\n— {gc.script_of(fam)}.py"); gc.run_block(fam)


— quark_amp.py
(0) the amplitudes
   scheme (T23): Q_u = 0.8489 in the working MSbar scheme (m_c(m_c)), 0.8316 with the pole charm mass; band 0.83-0.89; 5/6 = 0.8333 is the framed value
   leptons: Q=0.6667 r=1.4142 (=sqrt2=1.4142)
   down   : Q=0.7314 r=1.5455   (excess r-sqrt2=0.131)
   up     : Q=0.8489 r=1.7589   (excess r-sqrt2=0.345)

(1) DOWN: the excess is the GJ colour Clebsch (N_c=3) on the leptons
   GJ-dressed leptons (3 m_e, m_mu/3, m_tau): Q=0.7454, r=1.5724
   observed down                             : Q=0.7314, r=1.5455
[PASS] 1.gj_colour: colour (GJ, N_c=3) dressing shifts Q 2/3 -> 0.745 ~ observed Q_d=0.731 (2%): r_d-sqrt2 is the colour dressing

(2) UP: r_u ~ sqrt3 = 2 sin(pi/3), the cube-root/colour diagonal
   r_u = 1.7589  vs  sqrt3 = 2 sin(pi/3) = 1.7321  (1.5%)
   over m_u,m_c,m_t ranges: r_u = 1.759 [1.755,1.764] -- robust (m_u barely enters)
[PASS] 2.up_sqrt3: r_u=1.759 ~ sqrt3=1.732: the cube-root colour diagonal

(3) the diagonal over FRAMED RATIONALS: r^2

   p=53: N(1-i)=2=2 (four-fold, r^2=2, Q=2/3, sqrt2); N(1-omega)=3=3 (three-fold/colour, r^2=3, Q=5/6, sqrt3)


   p=89: N(1-i)=2=2 (four-fold, r^2=2, Q=2/3, sqrt2); N(1-omega)=3=3 (three-fold/colour, r^2=3, Q=5/6, sqrt3)
[PASS] 3.pattern: r^2 = N(1-zeta_n) = 2 (four-fold/colourless), 3 (three-fold/colour): integers in F_p

CONCLUSION: the quark amplitude excess r-sqrt2 is COLOUR, not a free parameter.
  Down: the GJ colour Clebsch (N_c=3, = M7) dresses the lepton spectrum, Q 2/3->0.745.
  Up:   r_u ~ sqrt3 = 2 sin(pi/3), the cube-root colour diagonal (cf lepton sqrt2=2sin(pi/4)).
  So r folds into the colour structure; r-sqrt2 is forced (down) / a sqrt3 lead (up).

checks: 6/6 pass
PASS the script's own tally: 6/6 checks
PASS [approx] Q_u = 0.8489 in the working scheme (0.849) and 0.8316 with the pole charm mass (0.832)
PASS [approx] the GJ-dressed lepton Q = 0.745 (0.745) against Q_d = 0.731 (0.731)
PASS [approx] r_u = 1.759 against √3 within 3 %
  [PASS] flv.quark_amp          MIXED  [28:B4, 28:C4, 28:X2, 28:P2] the quark amplitudes: the Georgi–Jarlskog colour dressing shifts Q from 2/3 to 0.

## The neutrino sector and leptonic CP
The seesaw of circulants; Q_ν = 2/3 solved on both orderings with the boundary branch selecting normal, Σm_ν = 59.1 meV; the magic matrix and maximal Jarlskog; the TM2 relation locking δ_CP to θ₂₃, its band and the joint-fit status.

| id | script | kind | claim | ledger rows |
|---|---|---|---|---|
| `flv.neutrino` | `neutrino.py` | MIXED | the seesaw of circulants circulant (symbolic, float-free); the signed Q_ν = 2/3 solved with the two measured Δm² on both orderings — normal (m₁ = 0.36, 8.62, 50.15 meV, Σ = 59.1 meV) and inverted (Σ = 102.3 meV) — the boundary branch selecting normal; the m₁ = 0 floor 58.8 meV at NuFIT 6.0; Σm_ν stable across the Δm² 1σ; r_ν = √2 reachable by the seesaw [approx] | 28:C11, 28:C11b, 28:P5, 28:P6, 28:X11 |
| `flv.pmns_cp` | `pmns_cp.py` | APPROX | the magic matrix trimaximal with maximal Jarlskog J = 1/(6√3) and the democratic column; over the TM2 family θ₂₃ = 45° ⇔ |δ_CP| = 90°; TM2 at the observed θ₁₃, θ₂₃ giving δ_CP ≈ −130° and sin²θ₁₂ ≈ 1/3 (numerical, numpy) | 28:C10, 28:P8, 28:X10 |
| `flv.tm2_jointfit` | `tm2_jointfit.py` | APPROX | the TM2 status: the trivial-singlet column protected (TM1 reducible), the exact-TM2 solar 0.341, the Cabibbo correction sweeping sin²θ₁₂ through 0.307, θ₁₃ = arcsin(sin θ_C/√2) ≈ 9.15°, the joint fit's parameter count (five for four observables, status not test), the TM2 relation against the direct PMNS evaluation, |δ_CP| = 130.4° at θ₂₃ = 49°, the band [50°, 156°] over the NuFIT 6.0 3σ range, the with-SK best fit as a live test (scipy) | 28:C9, 28:C10, 28:P8 |

Ledger rows witnessed: 28:C9, 28:C10, 28:C11, 28:C11b, 28:P5, 28:P6, 28:P8, 28:X10, 28:X11.

In [6]:
for fam in ["flv.neutrino", "flv.pmns_cp", "flv.tm2_jointfit"]:
    print(f"\n— {gc.script_of(fam)}.py"); gc.run_block(fam)


— neutrino.py
(1) EXACT: the seesaw of circulants is circulant  [float-free]


[PASS] 1.seesaw_circ: m_nu = -m_D M_R^-1 m_D^T is circulant for general circulant m_D,M_R (symbolic)
[PASS] 1.koide_form: => light neutrinos are Koide-form: sqrt(m_nu_k)=alpha+2|beta|cos(delta_nu+2pi k/3)

(2) PREDICTION [approx]: Q_nu=2/3 + Delta m^2 -> masses, ordering, Sum


   delta_nu=0.4778 rad; masses (meV): m1=0.36 m2=8.62 m3=50.15
   Sum m_nu = 59.1 meV ;  ordering NORMAL ;  lightest near boundary
[PASS] 2.normal: normal ordering, lightest m1=0.36 meV (near quarter-turn boundary)
[PASS] 2.sum_bound: Sum m_nu = 59.1 meV < 120 meV cosmology bound (Planck/DESI)
[PASS] 2.splittings: both measured Delta m^2 reproduced (the two inputs that fix the 2 free params)
[PASS] 2.signedQ_NO: signed Q_nu = 0.666666666667 on the normal root (one negative amplitude)


   INVERTED root: delta_nu=1.0420 rad; masses (meV): m3=2.93 m1=49.32 m2=50.07; Sum = 102.3 meV
[PASS] 2.inverted_exact: the signed circulant also solves inverted ordering exactly (signed Q_nu=0.666666666667); ordering is NOT fixed by Q_nu=2/3
[PASS] 2.boundary_branch: boundary branch: lightest amplitude -0.191 (normal) vs -0.414 (inverted) from the quarter-turn zero -> normal selected by the branch realisation (C11b), not by Q_nu

(3) [approx] robustness and seesaw consistency


   Dm^2 low: Sum m_nu = 58.7 meV


   Dm^2 high: Sum m_nu = 59.5 meV
[PASS] 3.robust: Sum m_nu = 58.7-59.5 meV across Delta m^2 1sigma (stable)
[PASS] 3.floor_nufit60: m1=0 floor at NuFIT 6.0 central values = 58.8 meV (common to every NO spectrum with m1=0)
[PASS] 3.seesaw_consistent: r_nu=sqrt2 achievable by the seesaw (min|r_nu-sqrt2|=9.5e-05): a consistency condition on the Dirac/Majorana phases, compatible with colourless Q_nu=2/3

SUMMARY: 11/11 checks PASS  (1 exact float-free + predictions/approx labelled)
PASS the script's own tally: 11/11 checks
PASS [approx] normal ordering: Σm_ν = 59.1 meV (the paper's 59.1) with m₁ = 0.36 meV
PASS [approx] inverted root: Σm_ν = 102.3 meV (the paper's 102.3)
PASS [approx] the m₁ = 0 floor at NuFIT 6.0 = 58.8 meV (58.8)
  [PASS] flv.neutrino           MIXED  [28:C11, 28:C11b, 28:P5, 28:P6, 28:X11] the seesaw of circulants circulant (symbolic, float-free); the signed Q_ν = 2/3 solved wit  --  15 checks, 5.8 s

— pmns_cp.py
(1) magic matrix: trimaximal, maximal Jarlskog
[PASS] 1

[PASS] 3.max90: all 378 TM2 points with theta23=45deg have |delta_CP|=90 (max dev 3.8 deg, grid-limited; exact at the magic point via J=Jmax). cos delta ∝ cot(2 theta23) = 0 at maximal

(4) TM2 at observed theta13, theta23 -> delta_CP


   sin^2 th12=0.341(obs .303)  sin^2 th13=0.0217(obs 0.0222)  sin^2 th23=0.569(obs 0.570)
   => delta_CP = -131 deg (drive-sign) ; global best fit (NuFIT NO) ~ -128 deg
[PASS] 4.match: TM2 at observed angles gives delta_CP=-131 deg, near maximal and matching the observed best fit ~-128 deg (-> exactly -90 as theta23 -> 45)
[PASS] 4.th12: TM2 predicts sin^2 th12=0.341 ~ 1/3 (obs 0.303; known mild TM2 tension)

(5) sign(delta_CP)<0 = drive orientation i_t=g^{-t} (same handedness as V-A) [structural]

SUMMARY: 6/6 checks PASS
PASS the script's own tally: 6/6 checks
PASS [approx] TM2 at the observed angles: δ_CP = −131° (the paper's ±124°/±130° band)
PASS [approx] J(F) = 0.09623 = 1/(6√3)
  [PASS] flv.pmns_cp            APPROX [28:C10, 28:P8, 28:X10] the magic matrix trimaximal with maximal Jarlskog J = 1/(6√3) and the democratic column; o  --  9 checks, 22.5 s

— tm2_jointfit.py
0  protected column: trivial C3 singlet (TM2) vs reducible doublet-sum (TM1)
  [OK] magic matrix trimaximal |F_

   A (CP=charged leptons): th13=9.1 th12=33.1 th23=43.8 dCP=100  chi2=76.1
   B (CP=neutrino delta_nu): th13=8.5 th12=33.7 th23=48.5 dCP=177  chi2=0.00
   parameter count: model A 3 parameters, model B 5 parameters, for 4 observables (theta_C imported):
   chi^2 ~ 0 in B is guaranteed and is not a test; the 'model A fails theta23' verdict is data-version dependent
   (NuFIT 6.0 with SK atmospheric data: theta23 = 43.3 deg, first octant).
  [OK] Model B (CP in neutrino delta_nu) reaches chi^2 ~ 0 with 5 parameters for 4 observables (no test)

4  the TM2 relation cos(dCP) = cot(2 th23) (1-2 s13^2)/(s13 sqrt(2-3 s13^2))
  [OK] relation = direct PMNS evaluation at (8.57, 49)
  [OK] delta_CP = +-130 deg at theta23 = 49 deg (the paper's earlier number)
   theta23= 41.0: cos dCP=+0.648  |dCP|=  49.6 deg
   theta23= 43.3: cos dCP=+0.274  |dCP|=  74.1 deg
   theta23= 45.0: cos dCP=+0.000  |dCP|=  90.0 deg
   theta23= 48.5: cos dCP=-0.566  |dCP|= 124.5 deg
   theta23= 50.6: cos dCP=-0.913  |dCP|

## The scales: transmutation, the coupling anchor, the fine-structure ledger
The overall scale as a transmutation exponent with y_t ≈ 1; the bare 1/4π against the unified continuum coupling; the α ledger — the 4π anchor, the electroweak decomposition, 3/8 at unification, the one-loop fermion sum, the 10³⁶ hierarchy.

| id | script | kind | claim | ledger rows |
|---|---|---|---|---|
| `flv.scale_a` | `scale_a.py` | APPROX | the overall scale as transmutation: y_t = √2 m_t/v ≈ 1 (the forced trigger); the one-loop descent of the quartic λ to the scale-invariant point (crossing zero below M_P); the hierarchy v/M_P = e^(−38.4); the lepton scale a² ≈ 314 MeV ≈ y_τ v [approx, dimensional] | 28:Z3 |
| `flv.coupling_anchor` | `coupling_anchor.py` | APPROX | the coupling anchor: α_bare⁻¹ = 4π against the unified continuum α_GUT⁻¹ ≈ 40 (the lattice–continuum matching Δ ≈ 27, the bare coupling stronger); α₁ = α₂ at ≈ 10¹³ GeV with the α₃ near-miss below 20 %; the couplings run to M_P [approx] | 28:Z4 |
| `flv.alpha_probe` | `alpha_probe.py` | APPROX | the fine-structure ledger: α_bare⁻¹ = 4π exact; α_EM⁻¹ = α₂⁻¹ + α_Y⁻¹ at M_Z; sin²θ_W → 3/8 at α₁ = α₂ (the derived generation content's b₁, b₂); 4π stronger than α_GUT so the bare-to-lab connection is a matching; the EM/gravity hierarchy α (m_P/m_p)² ≈ 1.2 × 10³⁶ [approx] | 28:A2, 28:Z4 |

Ledger rows witnessed: 28:A2, 28:Z3, 28:Z4.

In [7]:
for fam in ["flv.scale_a", "flv.coupling_anchor", "flv.alpha_probe"]:
    print(f"\n— {gc.script_of(fam)}.py"); gc.run_block(fam)


— scale_a.py
inputs at M_Z:  y_t = sqrt2 m_t/v = 0.9919 (~1, the forced trigger);  lambda(M_Z) = mH^2/2v^2 = 0.1294



  scale (GeV)      y_t       lambda
   9.12e+01    0.992    +0.1294
   2.43e+05    0.734    +0.0033
   6.47e+08    0.610    -0.0483
   1.72e+12    0.530    -0.0686
   4.59e+15    0.471    -0.0731
   1.22e+19    0.424    -0.0698

  lambda is driven DOWN by the (forced) top Yukawa: 0.13 at M_Z -> crosses 0 near mu ~ 3.4e+05 GeV (one-loop).
  -> the Higgs sits at the SCALE-INVARIANT (near-critical) point lambda~0 at high scale,
     exactly the boundary condition FRC scale-covariance forces.  (Two-loop places
     lambda~0 AND beta_lambda~0 right at M_P: Degrassi/Buttazzo near-criticality.)

  hierarchy v/M_P = 2.02e-17 = exp(-38.4): a transmutation exponent, not a tuning.
  exp(-c) with c~38 needs only an O(1) coupling: e.g. c=2pi/(b alpha), b~5, alpha~1/30 -> 38.

  the lepton scale a = (sum sqrt m)/3 = 17.72 MeV^1/2 ; a^2 = 314 MeV ~ lepton scale
  decomposition: a^2 ~ y_tau v, with v the transmutation EW scale (above) and
  y_tau = sqrt2 m_tau/v = 0.0102 the heaviest-lepton flavour n

## Summary
Writes `results.json` (one record per family: id, the ledger row(s) witnessed, script, kind, claim, PASS/FAIL, the micro-check count) and raises if any family failed.

In [8]:
ok = gc.summary(write=True)
failed = [r["id"] for r in gc.RESULTS if not r["ok"]]
assert ok, f"FAILED families: {failed}"
print(f"all {len(gc.RESULTS)} families pass ({len(gc.MICRO)} micro-checks); records in results.json")


SUMMARY: 16/16 checks passed (181 micro-checks)
all 16 families pass (181 micro-checks); records in results.json
